browser/ui/views/brave_actions/brave_actions_container.cc

```diff
@@ -0,0 +1,195 @@
+#include "ui/views/controls/separator.h"
+#include "ui/views/layout/box_layout.h"
+#include "ui/views/layout/grid_layout.h"
+#include "ui/views/view.h"
+#include "extensions/browser/extension_system.h"
+#include "extensions/browser/extension_registry.h"
+#include "extensions/browser/extension_registry_observer.h"
+#include "extensions/common/one_shot_event.h"
+
+BraveActionsContainer::BraveActionsContainer(Browser* browser, Profile* profile)
+  : views::View(),
+    browser_(browser),
+    extension_action_api_(extensions::ExtensionActionAPI::Get(profile)),
+    extension_registry_(extensions::ExtensionRegistry::Get(profile)),
+    extension_action_manager_(
+      extensions::ExtensionActionManager::Get(profile)
+    ),
+    extension_registry_observer_(this),
+    extension_action_observer_(this),
+    weak_ptr_factory_(this) {
+  // Handle when the extension system is ready
+  extensions::ExtensionSystem::Get(profile)->ready().Post(
+      FROM_HERE, base::Bind(&BraveActionsContainer::OnExtensionSystemReady,
+                            weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveActionsContainer::~BraveActionsContainer() {}
+
+void BraveActionsContainer::Init() {
+  // automatic layout
+  auto vertical_container_layout = std::make_unique<views::BoxLayout>(
+                                                views::BoxLayout::kHorizontal);
+  vertical_container_layout->set_main_axis_alignment(
+                                  views::BoxLayout::MAIN_AXIS_ALIGNMENT_CENTER);
+  vertical_container_layout->set_cross_axis_alignment(views::BoxLayout::CROSS_AXIS_ALIGNMENT_CENTER);
+  SetLayoutManager(std::move(vertical_container_layout));
+
+  // children
+  views::Separator* brave_button_separator_ = new views::Separator();
+  // TODO: theme color
+  brave_button_separator_->SetColor(SkColorSetRGB(0xb2, 0xb5, 0xb7));
+  brave_button_separator_->SetPreferredSize(gfx::Size(2,
+                            GetLayoutConstant(LOCATION_BAR_ICON_SIZE)));
+  AddChildView(brave_button_separator_);
+}
+
+void BraveActionsContainer::AddShields(
+                                const extensions::Extension* brave_extension) {
+  DCHECK(brave_extension);
+  VLOG(1) << "AddShields, was already loaded:" << (bool)shields_button_view_;
+  if (!shields_button_view_) {
+    // Create a ExtensionActionViewController for the extension
+    // Passing |nullptr| instead of ToolbarActionsBar since we do not require that
+    // logic.
+    // If we do require notifications when popups are open or closed,
+    // then we should inherit and pass |this| through.
+    shields_view_controller_ = new ShieldsActionViewController(brave_extension,
+        browser_, extension_action_manager_->GetExtensionAction(*brave_extension),
+                                                                        nullptr);
+    shields_button_view_ = new ToolbarActionView(
+                                                  shields_view_controller_, this);
```

Comment:
Let's call `shields_button_view_->set_owned_by_client()` because parent doesn't own button view.
With this, parent doesn't delete it by accident.

Reply:
I'm confused about this - BraveActionsContainer is the parent, since we add it as a child, right? If the Location Bar is deleted, it's ok to delete this button, no?

Reply:
Ah,,, I'm concerning there are two different delete path.(by tree view or by our `delete`).
That can cause some problem like double delete.

Reply:
oh i see, thanks. If I call that, does that mean I need to delete in destructor too?

Reply:
(FYI I was trying to avoid double delete by setting to `nullptr` but perhaps it can be deleted by tree view first?)

Reply:
Right. I worried that situation.
So, smart pointer can handle it.